README.md
<p align="center">
  <a href="https://github.com/TheGeeKing/TautulliPython2Trakt">
    <img src="image.png" alt="TautulliPython2Trakt's logo" width="80" height="80">
  </a>
  <h1 align="center">Tautulli Python 2 Trakt</h1>
</p>

## Table of Contents <!-- omit from toc -->

- [Description](#description)
- [What it can do](#what-it-can-do)
- [Requirements](#requirements)
- [Installation](#installation)
  - [Script Setup](#script-setup)
  - [Tautulli Setup](#tautulli-setup)
    - [Scrobbling](#scrobbling)
    - [Collection](#collection)
- [Usage](#usage)
- [More info](#more-info)
- [Similar Projects](#similar-projects)

## Description

Python script to scrobble what you watch, sync your collected movies and TV shows from [Tautulli](https://github.com/Tautulli/Tautulli) to [Trakt.tv](https://trakt.tv/).

## What it can do

- Track watched status of movies and TV shows
- Sync collected movies and TV shows to Trakt.tv

## Requirements

- Tautulli 2.12.5 or higher
- Python 3.11 or higher
- install requirements.txt
  1. `pip install -r requirements.txt`

## Installation

### Script Setup

Download the latest release from [here](https://github.com/TheGeeKing/TautulliPython2Trakt/releases), unzip it and place all files in a folder.

Create a new [application](https://trakt.tv/oauth/applications) and add the following settings:

**Name:** `TautulliPython2Trakt` \
**Redirect uri:** `urn:ietf:wg:oauth:2.0:oob` \
**Permissions:** `/scrobble`

Run the script:

```bash
python TautulliPython2Trakt.py -setup
```

Follow the setup steps.

If you want to collect your movies and TV shows, you need to do the Plex Media Server setup! You might also want to add credentials/token to access your Plex Media Server as this can help to scrobble the correct episode to Trakt.tv if you don't use the TMDB database for sorting your shows. [How to find my token ?](https://support.plex.tv/articles/204059436-finding-an-authentication-token-x-plex-token/)

### Tautulli Setup

#### Scrobbling

1. Go to Tautulli > `Settings` > `Notification Agents` > `Add a new notification agent` > `Script`
2. Set the `Script Folder` to the folder where you placed the script, then select the script `TautulliPython2Trakt.py` in the `Script File` field. Put the `Script Timeout` to `0`.
3. In the `Triggers` section, select:
   1. `Playback Start`
   2. `Playback Stop`
   3. `Playback Pause`
   4. `Playback Resume`
   5. `Watched`
4. Put conditions if you want to, like username, media type, etc.
5. In the `Arguments` tab, put the following arguments:
   1. Playback Start / Playback Resume: `pythonw -m {media_type} -s "{show_name}" -M "{title}" -y "{year}" -t "{thetvdb_id}" -i "{imdb_id}" -r {rating_key} -S {season_num} -E {episode_num} -P {progress_percent} -a start -PlexUser {username}`
   2. Playback Stop / Watched: `pythonw -m {media_type} -s "{show_name}" -M "{title}" -y "{year}" -t "{thetvdb_id}" -i "{imdb_id}" -r {rating_key} -S {season_num} -E {episode_num} -P {progress_percent} -a stop -PlexUser {username}`
   3. Playback Pause: `pythonw -m {media_type} -s "{show_name}" -M "{title}" -y "{year}" -t "{thetvdb_id}" -i "{imdb_id}" -r {rating_key} -S {season_num} -E {episode_num} -P {progress_percent} -a pause -PlexUser {username}`

#### Collection

1. Go to Tautulli > `Settings` > `Notification Agents` > `Add a new notification agent` > `Script`.
2. Set the `Script Folder` to the folder where you placed the script, then select the script `TautulliPython2Trakt.py` in the `Script File` field. Put the `Script Timeout` to `0`.
3. In the `Triggers` section, select `Recently Added`.
4. Put conditions if you want to, like media type, etc.
5. In the `Arguments` tab, put the following argument:
   1. Recently Added: `<movie>pythonw -c movies -PlexUser %OWNER%</movie><episode>pythonw -c episodes -PlexUser %OWNER%</episode><season>pythonw -c episodes -PlexUser %OWNER%</season><show>pythonw -c episodes -PlexUser %OWNER%</show>`

## Usage

```
-h                  Help
-setup              Setup aplication
-reset              Reset settings and revoke token
-refreshToken       Refreshes the Trakt token
-add                Add a new Plex user
-remove             Remove a Plex user

------------------ Trakt Scrobbling ------------------
-m                  Media type (movie, show, episode)
-a                  Action (start, pause, stop)
-s                  Name of the TV Series
-M                  Name of the Moviename
-y                  Year of the movie/TV Show
-S                  Season number
-E                  Episode number
-t                  TVDB ID
-i                  IMDB ID
-r                  ratingKey (Plex Server internal content id)
-P                  Percentage progress (Ex: 10.0)
-PlexUser           The Plex username

------------------ Trakt Collection ------------------
-c                  Media type (movies, episodes)
-PlexUser           The Plex username (check 'Syncing behavior' in section 'More info' in the README.md file)
```

## More info

Default scrobbler behavior is for:

- Movies, the data from Tautulli is sent directly to Trakt.
- Episodes:
  - If your Plex Media Server is connected, we get the ratingkey from the data sent by Tautulli. We make a database filled with ratingkey paired to ids. We search for the ids linked to the ratingkey in the database. We send the ids to Trakt. Trakt.tv uses TMDB database, so sending basic info like season and episode number can mismatch with your plex configuration. This way we ensure that the episode is scrobbled to the correct one on the Trakt end.
  - If you are not connected to your Plex Media Server, we send the data from Tautulli directly to Trakt.

Syncing behavior:

- Based on the -c argument, we either sync movies or episodes. It is syncing your entire collection, not just the recently added, so it might take some time. If it takes way too much time, open an issue and I might add/find a way to only sync the recently added content.
- Based on the -PlexUser argument:
  - (default behavior) `%OWNER%`, we sync the collections to the owner Trakt account.
  - If a user is specified, we sync the collections to the specified user Trakt account. ⚠️ **It will sync like if it was the owner, so even if the user has not access to the library where the content was added.** You can also use a list: `"[username1, username2]"`, typo is very important.
  - If `%ALL%` is specified, we sync the collections to all the users Trakt account. It will check if the users have access to the content before adding it to their collection.

## Similar Projects

Inspired from: https://github.com/frugglehost/TautulliBatch2Trakt

- https://github.com/JvSomeren/tautulli-watched-sync
- https://github.com/xanderstrike/goplaxt
- https://github.com/gazpachoking/trex
- https://github.com/dabiggm0e/plextrakt
- https://github.com/trakt/Plex-Trakt-Scrobbler
- https://github.com/Generator/tautulli2trakt


CHANGELOG.md
# Changelog <!-- omit from toc -->

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

- Refactor code for requests.
- Use a proper logging module.
- Use proper argument conventions (-h, --help, case) and maybe an argument parser.
- Maybe use a different way to `sys.exit()` known errors.

## [1.1.0] - 2023-08-07

### Added

- This CHANGELOG.md file.
- GitHub Templates for Bug Reports and Feature Requests.
- Syncing behavior section in the README.md file.
- More info for the `-PlexUser` argument in the README.md file and when using the `-h` argument.

### Fixed

- Sync collections now correctly works.
- Sentences in README.md file now go correctly to the next line and grammar mistakes.
- Syncing collections now correctly works.
- Now correctly log the arguments passed before calling `subprocess.check_output()`.
- Fixed some puctuation mistakes.

### Changed

- Argument for the recently added in the README.md file.
- Syncing collection can now either find the owner username, sync all users found in the env, or sync a list of users.
- Sorted imports.
- Now gets `HEADERS` from a function in utilities.py for less redundancy.

### Removed

- Unused import os in scrobble.py.
- Unused `arguments_string` variable in the TautulliPython2Trakt.py file.

## [1.0.0] - 2023-05-27

### Added

- Initial release of TautulliPython2Trakt v1.0.0.

[unreleased]: https://github.com/TheGeeKing/TautulliPython2Trakt/compare/v1.1.0...main
[1.1.0]: https://github.com/TheGeeKing/TautulliPython2Trakt/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/TheGeeKing/TautulliPython2Trakt/releases/tag/v1.0.0


sql.py
import os
import sqlite3


class SQL(sqlite3.Cursor):
    def __init__(self, db: str = "database.db", path: str = ""):
        self.con = sqlite3.connect(os.path.join(path, db), timeout=5)
        self.con.row_factory = (
            sqlite3.Row
        )  # So we can use column names as keys, cause less issues if we later add a column with index values
        sqlite3.Cursor.__init__(self, self.con)
        self.cur = self.con.cursor()

    def execute(self, command: str, *args):
        with self.con:
            if args:
                tmp = self.con.execute(command, args[0])
            else:
                tmp = self.con.execute(command)
            self.con.commit()  # sqlite3.Connection automatically commits
        # tmp = self.cur.execute(command)
        # self.con.commit()
        return tmp

    def commit(self):
        return self.con.commit()

    def add_simple(self, table: str, **kwargs) -> sqlite3.Cursor:
        """Simply add values to the table, no need to specify columns

        Args:
            table (str): the table to add to
            **kwargs: the values to add to the table

        Returns:
            Cursor: the cursor of the db
        """
        with self.con:
            # column_names = self.execute(f"PRAGMA table_info({table})").fetchall()
            tmp = self.con.execute(
                f"INSERT INTO {table} ({', '.join(kwargs)}) VALUES ({', '.join(['?'] * len(kwargs))})",
                tuple(kwargs.values()),
            )
            self.con.commit()
        return tmp

    def add(self, table: str, data: tuple[tuple[str, type]]) -> sqlite3.Cursor:
        """Insert data into the table with tuple as (column, value)

        Args:
            table (str): the table to add to
            data (tuple): ((column, value), (column, value))

        Returns:
            sqlite3.Cursor: the cursor of the db
        """
        with self.con:
            tmp = self.con.execute(
                f"INSERT INTO {table} ({', '.join([i[0] for i in data])}) VALUES ({', '.join(['?'] * len(data))})",
                tuple([i[1] for i in data]),
            )
            self.con.commit()
        return tmp

    def create_table(self, table: str, data: tuple[tuple] | str) -> sqlite3.Cursor:
        if isinstance(data, str):
            with self.con:
                if "PRIMARY KEY" not in data:
                    exec_ = f"CREATE TABLE IF NOT EXISTS {table} (id INTEGER PRIMARY KEY AUTOINCREMENT, {data})"
                else:
                    exec_ = f"CREATE TABLE IF NOT EXISTS {table} ({data})"
                tmp = self.con.execute(exec_)
                self.con.commit()
            return tmp
        types = []
        for d in data:
            if d[1] == int:
                types.append(f"{d[0]} INTEGER")
            elif d[1] == str:
                types.append(f"{d[0]} TEXT")
            elif d[1] == float:
                types.append(f"{d[0]} REAL")
            elif d[1] == bool:
                types.append(f"{d[0]} BOOLEAN")
            else:
                # raise TypeError(f"Type {d[1]} is not supported")
                types.append(f"{d[0]} {d[1]}")
        types = ", ".join(types)
        if "PRIMARY KEY" not in types:
            exec_ = f"CREATE TABLE IF NOT EXISTS {table} (id INTEGER PRIMARY KEY AUTOINCREMENT, {types})"
        else:
            exec_ = f"CREATE TABLE IF NOT EXISTS {table} ({types})"
        with self.con:
            self.con.execute(exec_)

    def add_safe(self, table: str, data: tuple[tuple]) -> sqlite3.Cursor:
        """If the table does not exist, create it. Insert data into the table with tuple as (column, value)

        Args:
            table (str): the table to add to
            data (tuple): ((column, value), (column, value))

        Returns:
            sqlite3.Cursor: the cursor of the db
        """
        with self.con:
            if not self.table_exists(table):
                tmp = self.create_table(table, data)
            tmp = self.add(table, data)
            self.con.commit()
        return tmp

    def table_exists(self, table: str) -> bool:
        """Check if a table exists

        Args:
            table (str): the table to check

        Returns:
            bool: if the table exists
        """
        tmp = self.execute(
            "SELECT name FROM sqlite_master WHERE type='table' AND name=?", (table,)
        )
        return bool(tmp.fetchall())

    # def insert(conn, command: str):

    #     with conn.cursor() as c:
    #         c.execute(...)
    #     conn.commit()

    # def execute_soft(self, command: str):
    #     return self.cur.execute(command)

    def executemany(self, command: str, args: list[tuple]):
        tmp = self.cur.executemany(command, args)
        self.con.commit()
        return tmp


get_ids.py
import json
import os
import sys
import threading

from dotenv import load_dotenv
from plexapi.library import MovieSection, ShowSection

from sql import SQL
from utilities import get_plex_server, guids_to_ids

load_dotenv()

ARGS = sys.argv[1:]

plex = get_plex_server()

ALL_SECTIONS = plex.library.sections()
MOVIES_SECTIONS = [
    section for section in ALL_SECTIONS if isinstance(section, MovieSection)
]
SHOWS_SECTIONS = [
    section for section in ALL_SECTIONS if isinstance(section, ShowSection)
]


def get_json_ids(guids):
    """
    The function `get_json_ids` takes a list of objects with `id` attributes and returns a dictionary
    where the keys are the parts of the `id` before `://` and the values are the parts after `://`,
    converted to integers if the key is either "tmdb" or "tvdb".

    :param guids: A list of objects containing ids
    :return: a dictionary containing the IDs extracted from the given list of guids.
    """
    ids = {}
    for obj in guids:
        # add the id to the dict, format is taking before :// as key and after as value
        ids[f"{obj.id.split('://')[0]}"] = (
            int(obj.id.split("://")[1])
            if f"{obj.id.split('://')[0]}" in ["tmdb", "tvdb"]
            else obj.id.split("://")[1]
        )
    return ids


def _create_database():
    db = SQL(db="ratingkeys.db")
    db.create_table("ratingkeys", (("ratingkey", "INTEGER UNIQUE"), ("ids", "JSON")))


def setup_database():
    if not does_database_exist("ratingkeys.db"):
        _create_database()
    db = SQL(db="ratingkeys.db")

    to_add: list[tuple] = []
    for lib in ALL_SECTIONS:
        search = lib.searchEpisodes() if isinstance(lib, ShowSection) else lib.search()
        for content in search:
            ids = guids_to_ids(content.guids)
            if ids:
                to_add.append((content.ratingKey, json.dumps(ids)))
    db.executemany("INSERT INTO ratingkeys (ratingkey, ids) VALUES (?, ?)", to_add)


def does_database_exist(name: str):
    return os.path.exists(name if name.endswith(".db") else f"{name}.db")


def is_database_empty(table: str):
    if not does_database_exist("ratingkeys.db"):
        return True
    db = SQL(db="ratingkeys.db")
    res = db.execute(f"SELECT * FROM {table}").fetchone()
    return res is None


def update_database():
    """
    The function `update_database` updates the `ids` column in the `ratingkeys` table of a database with
    the JSON representation of the content's guids, for each content in the search results.
    """
    db = SQL(db="ratingkeys.db")
    to_update = []
    for lib in ALL_SECTIONS:
        search = lib.searchEpisodes() if isinstance(lib, ShowSection) else lib.search()
        for content in search:
            ids = json.dumps(guids_to_ids(content.guids))
            to_update.append(({"ratingkey": content.ratingKey, "ids": ids}))
    db.executemany(
        "UPDATE ratingkeys SET ids = :ids WHERE ratingkey = :ratingkey AND ids != :ids",
        to_update,
    )


def _raw_search_for_media_ratingkey_database(ratingkey: int):
    if is_database_empty("ratingkeys"):
        setup_database()
    db = SQL(db="ratingkeys.db")
    res = db.execute(
        "SELECT * FROM ratingkeys WHERE ratingkey = ? LIMIT 1", (ratingkey,)
    ).fetchone()
    return res["ids"] if res else None


def verify_ratingkey(media, ratingkey: int):
    return None if media["ratingKey"] != ratingkey else media


def _raw_search_for_media_ratingkey(media_type: ["movie", "episode"], ratingkey: int):
    SECTIONS = MOVIES_SECTIONS if media_type == "movie" else SHOWS_SECTIONS
    for lib in SECTIONS:
        search = lib.searchEpisodes() if media_type == "episode" else lib.search()
        for content in search:
            if content.__dict__["ratingKey"] == ratingkey:
                return content
    # return verify_ratingkey(plex.fetchItem(ratingkey), ratingkey) /!\ Non working module is trying to but -> 'Failed to parse: http://localhost:3240013954' the ratingkey is merged with the port. otherwise might be a better solution to directly ask the plex server with plex.fetchItem


def search_for_media_ratingkey(
    media_type: ["movie", "episode"], ratingkey: int, database_search: bool = True
):
    if database_search:
        ids = _raw_search_for_media_ratingkey_database(ratingkey)
        if ids:
            return json.loads(ids)
    obj = _raw_search_for_media_ratingkey(media_type, ratingkey)
    if obj:
        # If we are here, it means that the database is outdated
        thread = threading.Thread(target=update_database)
        thread.start()
        return get_json_ids(obj.__dict__["guids"])
    return None


scrobble.py
import json
import sys
import time

import requests
from dotenv import load_dotenv

from get_ids import search_for_media_ratingkey
from utilities import get_from_env, log

load_dotenv()

args = sys.argv[1:]
NUMBER_OF_ARGS = len(args)


log(f"args:\n{args}\n")
log(f"NUMBER_OF_ARGS: {NUMBER_OF_ARGS}\n")

i = 0
while i < NUMBER_OF_ARGS:
    current_arg = args[i]
    if (
        current_arg.startswith("-")
        and i + 1 < NUMBER_OF_ARGS
        and not args[i + 1].startswith("-")
    ):  # if it isn't a show -s is empty followed by -M, so we go to the next arg -M otherwise we would go to the title string
        pass
    else:
        i += 1
        continue
    # log(f"{i}: {current_arg}")
    arg = args[i + 1]
    match current_arg:
        case "-m":
            MEDIA_TYPE = arg
        case "-s":
            SHOW_NAME = arg
        case "-M":
            MOVIE_NAME = arg
        case "-y":
            MEDIA_YEAR = arg
        case "-t":
            TVDB_ID = int(arg)
        case "-i":
            IMDB_ID = arg
        case "-r":
            RATINGKEY = arg
        case "-S":
            SEASON = int(arg)
        case "-E":
            EPISODE = int(arg)
        case "-P":
            PROGRESS = float(arg)
        case "-a":
            ACTION = arg
        case "-PlexUser":
            PLEX_USER = arg
    i += 2

if "MEDIA_YEAR" not in locals():
    MEDIA_YEAR = 0

if MEDIA_YEAR == "":
    MEDIA_YEAR = 0

body = ""
log(f"media_type: {MEDIA_TYPE}")
if MEDIA_TYPE == "movie":
    body = {
        "movie": {
            "title": f"{MOVIE_NAME}",
            "year": f"{MEDIA_YEAR}",
            "ids": {"imdb": f"{IMDB_ID}"},
        }
    }
elif MEDIA_TYPE in ["show", "episode"]:
    if get_from_env("PLEX_BASEURL") and search_for_media_ratingkey(
        MEDIA_TYPE, RATINGKEY
    ):  # this is slower, but it's more reliable if season/episode database is not the same one as Trakt.tv
        ids = search_for_media_ratingkey(MEDIA_TYPE, RATINGKEY)
        body = {"episode": {"ids": ids}}
    else:
        body = {
            "show": {
                "title": f"{SHOW_NAME}",
                "year": MEDIA_YEAR,
                "ids": {"tvdb": TVDB_ID},
            },
            "episode": {"season": SEASON, "number": EPISODE},
        }

else:
    log("Invalid media type")
    sys.exit(1)

body["progress"] = PROGRESS
body["app_version"] = "1.0"
body["app_date"] = time.strftime("%Y-%m-%d")

URI = f"https://api.trakt.tv/scrobble/{ACTION}"

HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {get_from_env('access_token', PLEX_USER)}",
    "trakt-api-version": "2",
    "trakt-api-key": get_from_env("client_id", PLEX_USER),
}

body = json.dumps(body)

log(body)
log(URI)

r = requests.post(URI, headers=HEADERS, data=body)
log(f"{r.status_code} | {r.text}")
if r.status_code == 404:
    if MEDIA_TYPE == "movie":
        ids = search_for_media_ratingkey(MEDIA_TYPE, RATINGKEY)
        if ids:
            body = {
                "movie": {
                    "ids": ids,
                }
            }
            body["progress"] = PROGRESS
            body["app_version"] = "1.0"
            body["app_date"] = time.strftime("%Y-%m-%d")
            body = json.dumps(body)
            r = requests.post(URI, headers=HEADERS, data=body)
        else:
            msg = "404, movie or episode not found on Trakt.tv. If this is incorrect, please report it on github."
            log(msg)
            print(msg)
            sys.exit(1)
while r.status_code == 429:
    time.sleep(5)
    r = requests.post(URI, headers=HEADERS, data=body)
    log(f"{r.status_code} | {r.text}")
print(r.status_code)
sys.exit(0)


TautulliPython2Trakt.py
import json
import os
import subprocess
import sys
import time

import requests
from dotenv import load_dotenv, set_key
from plexapi.myplex import MyPlexAccount

from utilities import ProgressBar, get_from_env, log

args = sys.argv[1:]

if not args:
    print("No arguments given! Use -h for help.")
    sys.exit(1)
log(f"initial startup args:\n{args}\n")

if args[0] == "-h":
    print(
        """
# <- not yet implemented
-h                  Help
-setup              Setup aplication
-reset              Reset settings and revoke token
-refreshToken       Refreshes the Trakt token
-add                Add a new Plex user
-remove             Remove a Plex user

------------------ Trakt Scrobbling ------------------
-m                  Media type (movie, show, episode)
-a                  Action (start, pause, stop)
-s                  Name of the TV Series
-M                  Name of the Moviename
-y                  Year of the movie/TV Show
-S                  Season number
-E                  Episode number
-t                  TVDB ID
-i                  IMDB ID
-r                  ratingKey (Plex Server internal content id)
-P                  Percentage progress (Ex: 10.0)
-PlexUser           The Plex username

------------------ Trakt Collection ------------------
-c                  Media type (movies, episodes)
-PlexUser           The Plex username (check 'Syncing behavior' in section 'More info' in the README.md file)
#-A                  Collection action (add, remove)

          """
    )
    sys.exit(0)


def _add_user_env(user):
    if not load_dotenv():
        with open(".env", "w", encoding="utf-8") as f:
            f.write("DATA='{}'")
    load_dotenv()
    data = json.loads(get_from_env("DATA"))
    data.update(user)
    set_key(".env", "DATA", json.dumps(data))


def _setup_user_process():
    plex_user = input("Enter the Plex username: ")
    trakt_client_id = input("Enter Trakt.tv 'Client ID': ")
    trakt_client_secret = input("Enter Trakt.tv 'Client Secret': ")
    if not all([plex_user, trakt_client_id, trakt_client_secret]):  # if empty
        print("One or more fields are empty!")
        sys.exit(1)

    req = requests.post(
        "https://api.trakt.tv/oauth/device/code", data={"client_id": trakt_client_id}
    )
    print(
        req.json()
    )  # {'device_code': '6764ad68383eca00b3e3a960b0cf02dcbc055dab2452f8dd583e786683578e9c', 'user_code': 'FB222899', 'verification_url': 'https://trakt.tv/activate', 'expires_in': 600, 'interval': 5}

    device_code = req.json()["device_code"]
    user_code = req.json()["user_code"]
    expires_in = req.json()["expires_in"]

    print(
        f"""
Autorize the aplication.
1. Open the URL https://trakt.tv/activate
2. Copy the temp code {user_code} (Note it is already in the windows clipboard)
3. Accept Web prompts.

This Screen will auto refresh untill the token is accepted.
There are {expires_in} seconds untill the code {user_code} expires.
"""
    )
    os.system(f"start {req.json()['verification_url']}")
    os.system(f"echo {user_code}| clip")

    status_code = 0
    progress_bar = ProgressBar("Granting access")
    progress_bar.start()
    while status_code != 200 and status_code != 409 and expires_in > 0:
        time.sleep(5)
        expires_in -= 5
        body = {
            "code": device_code,
            "client_id": trakt_client_id,
            "client_secret": trakt_client_secret,
        }
        req = requests.post(
            "https://api.trakt.tv/oauth/device/token", data=body
        )  # {'access_token': 'xxx', 'token_type': 'bearer', 'expires_in': 7776000, 'refresh_token': 'xxx', 'scope': 'public', 'created_at': 1690327181}
        status_code = req.status_code

        expires_in -= 5
        progress_bar.update((600 - expires_in) / 600 * 100)
    progress_bar.end()

    user = {
        plex_user: {"client_id": trakt_client_id, "client_secret": trakt_client_secret}
    }
    for key, value in req.json().items():
        user[plex_user][key] = value

    # set_key(".env", "DATA", json.dumps(user))
    return user


def setup_plex():
    if not load_dotenv():
        print("Try to run -setup first!")
        sys.exit(1)
    # print("This part assume that the Plex Media Server is running on the same machine as this script for now.")
    print("Recommended to run Plex Media Server on same machine as this script.")
    print(
        """
Next step you have 2 choices:
1. If your Plex Media Server is running on the same machine as this script, you can use local login with plex url and token.
2. If your Plex Media Server is running on a different machine or you want to use remote login, you will need to provide username and password. The script will then generate a token for you and store it in the .env file. Your username and password will not be stored. You can check the ".env" file after or read the code.

case 1 is recommended if you are running the script on the same machine as the Plex Media Server or you have some IT knowledges.
case 2 is recommended if you are running the script on a different machine than the Plex Media Server or you don't have any IT knowledges and have your credentials with you.
"""
    )
    choice = int(input("Enter your choice (1/2): "))
    if choice == 1:
        PLEX_BASEURL = input("Enter Plex Base URL (default: http://localhost:32400): ")
        PLEX_TOKEN = input(
            "Enter Plex Token (https://support.plex.tv/articles/204059436-finding-an-authentication-token-x-plex-token/): "
        )
    elif choice == 2:
        PLEX_USERNAME = input("Enter your Plex Username: ")
        PLEX_PASSWORD = input(
            "Enter your Plex Password (if 2fa is enabled add the code at the end e.g. password123456): "
        )
        PLEX_BASEURL = input(
            "Enter the Plex Media Server name (name on the top left when view more): "
        )
        if PLEX_USERNAME and PLEX_PASSWORD:
            account = MyPlexAccount(PLEX_USERNAME, PLEX_PASSWORD)
            PLEX_TOKEN = account.authenticationToken
        else:
            print("Invalid credentials!")
            sys.exit(1)
    else:
        print("Invalid choice!")
        sys.exit(1)

    set_key(".env", "PLEX_BASEURL", PLEX_BASEURL or "http://localhost:32400")
    set_key(".env", "PLEX_TOKEN", PLEX_TOKEN)
    print(
        """
Your Plex Media Server credentials are stored in the .env file.
If you modify your password, the token might become invalid.
You will need to run -setup again to update the token or manually editing it in the .env file.
"""
    )


def add_user_process():
    user = _setup_user_process()
    _add_user_env(user)


if args[0] == "-setup":
    if (
        not load_dotenv()
        or not get_from_env("DATA")
        or not json.loads(get_from_env("DATA"))
    ):
        add_user_process()
        print(
            """
Some of your shows might use a different season/episode numbering system than Trakt.tv.
The fallback method to avoid any issues requires access to your Plex Media Server to retrieve the needed data to point to the correct episode.
Syncing your Plex collection to Trakt.tv also requires access to your Plex Media Server.
"""
        )
        setup_plex_ = input(
            "Do you want to setup access to your Plex Media Server ? (y/n): "
        )
        if setup_plex_.lower() == "y":
            setup_plex()
        print("Setup completed!")
        sys.exit(0)
    else:
        print("The .env file already exists!")
        print(
            "If you want to reset the settings, use -reset. If you want to add a new user, use -add. For more help, use -h"
        )
        sys.exit(1)

if args[0] == "-reset":
    if load_dotenv() and get_from_env("DATA"):
        os.remove(".env")
        if os.path.exists("ratingkeys.db"):
            os.remove("ratingkeys.db")
            print(
                "The .env file and the ratingkeys.db file have been deleted! You can now run -setup"
            )
        else:
            print("The .env file has been deleted! You can now run -setup")
    else:
        print("Haha, there is no data, you should first run -setup!")
    sys.exit(0)


def _refresh_token(user: dict, data: dict):
    print(data, type(data))
    print(user, type(user))

    body = {
        "refresh_token": user["refresh_token"],
        "client_id": user["client_id"],
        "client_secret": user["client_secret"],
        "redirect_uri": "urn:ietf:wg:oauth:2.0:oob",
        "grant_type": "refresh_token",
    }
    req = requests.post(
        "https://api.trakt.tv/oauth/token", data=body
    )  # {"access_token": "xxx", "token_type": "Bearer", "expires_in": 7775999, "refresh_token": "xxx", "scope": "public", "created_at": 1690376320}

    for key, value in req.json().items():
        user[key] = value
    set_key(".env", "DATA", json.dumps(data))


def refresh_token(plex_user=None):
    data = json.loads(get_from_env("DATA"))
    # user: { "created_at": 123, "expires_in": 12}
    # if created_at + expires_in < 1 month till the date, then refresh the token
    if plex_user:
        user = data[plex_user]
        _refresh_token(user, data)
    else:
        for user in data:
            _refresh_token(data[user], data)


def less_month_expiration_token_users(data):
    """
    The function `less_month_expiration_token_users` returns a list of users whose tokens are expiring
    within one month.

    :param data: The `data` parameter is a list of dictionaries, where each dictionary represents a
    user. Each user dictionary should have the following keys:
    :return: a list of users whose tokens are expiring within one month.
    """
    users = []
    for key, value in data.items():
        user = value
        # using time to get the 1 month in seconds
        time_trigger = 60 * 60 * 24 * 30
        if (
            user["created_at"] + user["expires_in"] < time.time() + time_trigger
        ):  #  and user["created_at"] + user["expires_in"] > time.time() - time_trigger #* see comment below
            users.append(key)
        # elif (user["created_at"] + user["expires_in"]) - time.time() < 0:
        #     log(f"User {user['username']} token has expired ! You need to -setup") #* not sure if the refresh_token work indefinitely, in case not, this will be usefull
    return users


if args[0] == "-refreshToken":
    if load_dotenv() and get_from_env("DATA"):
        refresh_token()
        sys.exit(0)
    else:
        print("Haha, there is no data, you should first run -setup !")
        sys.exit(1)


if load_dotenv() and get_from_env("DATA"):
    users = less_month_expiration_token_users(json.loads(get_from_env("DATA")))
    if users:
        for user in users:
            refresh_token(user)
            print(f"User {user}'s token has been refreshed!")
            if len(users) > 1:
                time.sleep(3)  # to be sure not to get a 429 (rate limit)


if args[0] == "-add":
    add_user_process()
    sys.exit(0)

if args[0] == "-remove":
    data = json.loads(get_from_env("DATA"))

    if load_dotenv() and get_from_env("DATA"):
        data = json.loads(get_from_env("DATA"))
        plex_user = input("Enter the Plex username: ")
        if plex_user in data:
            del data[plex_user]
            set_key(".env", "DATA", json.dumps(data))
            print(f"User {plex_user} has been removed!")
            sys.exit(0)
        else:
            print(f"User {plex_user} doesn't exist!")
    else:
        print("Haha, there is no data, you should first run -setup !")
    sys.exit(1)


if not load_dotenv():
    raise FileNotFoundError("Couldn't load .env file! Run -setup to create it.")

NUMBER_OF_ARGS = len(args)

arguments_list = []
if args[0] == "-m":
    SCROBBLE = True
    i = 0
    while i < NUMBER_OF_ARGS:
        current_arg = args[i]
        if (
            current_arg.startswith("-")
            and i + 1 < NUMBER_OF_ARGS
            and not args[i + 1].startswith("-")
        ):  # if it isn't a show -s is empty followed by -M, so we go to the next arg -M otherwise we would go to the title string
            next_arg = args[i + 1] if i + 1 < NUMBER_OF_ARGS else ""
            if " " in next_arg:
                next_arg = f'"{next_arg}"'
            arguments_list += [current_arg, next_arg]
            i += 2
        else:
            i += 1
else:
    SCROBBLE = False

log(f"ARGUMENTS: {args}")

log(f"SCROBBLE: {SCROBBLE}")
if SCROBBLE:
    proc = subprocess.check_output(["py", "scrobble.py"] + args, shell=True, text=True)
else:
    proc = subprocess.check_output(
        ["py", "sync_collections.py"] + args, shell=True, text=True
    )

log(proc)
print(proc)
